src/Workspaces/Core/Portable/Workspace/Workspace_Events.cs

```diff
@@ -89,20 +90,31 @@ protected Task RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind kind, Solutio
             RaiseEventForHandlers(ev, sender: this, args, FunctionId.Workspace_EventsImmediate);
         }
 
+        var eventHandlerTasks = new List<Task>();
+
         ev = GetEventHandlers<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
         if (ev.HasHandlers)
         {
             args ??= new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
-            return this.ScheduleTask(() =>
+            var syncEventHandlersTask = this.ScheduleTask(() =>
             {
                 RaiseEventForHandlers(ev, sender: this, args, FunctionId.Workspace_Events);
             }, WorkspaceChangeEventName);
+
+            eventHandlerTasks.Add(syncEventHandlersTask);
         }
-        else
+
+        var asyncEv = _asyncEventMap.GetEventHandlerSet<WorkspaceChangeEventArgs>(WorkspaceChangeEventName);
+        if (asyncEv.HasHandlers)
         {
-            return Task.CompletedTask;
+            args ??= new WorkspaceChangeEventArgs(kind, oldSolution, newSolution, projectId, documentId);
+            var asyncEventHandlersTask = this.ScheduleBackgroundTask(() => RaiseEventForAsyncHandlersAsync(asyncEv, args, FunctionId.Workspace_Events));
+
+            eventHandlerTasks.Add(asyncEventHandlersTask);
         }
 
+        return Task.WhenAll(eventHandlerTasks);
```

Comment:
interesting.  i was not expecting this.  i woudl not have thought that raising the event would have us wait on the receiver completing their work.

I'd be moer of a fan of the event just firing on the BG.  and potentially telling receivers: you need to be synchronous.  if you need async, add an item to your own async BG queue.

Reply:
If this is the one I'm thinking of, it's a strange method in that it returns a Task representing the completion of the work...which all of the callers then ignore. I'd absolutely say we should change this signature if it wasn't the fact that's a breaking change...